### PR TITLE
CRM: Introduce Jetpack Autoloader and declare Dompdf in Composer

### PR DESCRIPTION
**Pull request by kallehauge**
We used to have a "hardcoded" version of `Dompdf` in our repository which doesn't play nice with some of Jetpack Monorepo's `.gitignore` rules.

This PR aim to correct that by implementing `Dompdf` via Composer instead and autoload the package when needed.

## Proposed changes:

* Introduce `Jetpack Autoloader`.
* Define Dompdf dependency in composer.
* Refactor existing code to prevent `require()` attempts for the old location.
* Add a deprecated tag to `ZeroBSCRM::checkDompdfVersion`.
  * It shouldn't be necessary to do so due to the `ZeroBSCRM` class being `final` but we'll remove it with a refactor where we remove the entire library system.
* Remove all files in `projects/plugins/crm/includes/lib/`  since we no longer use them.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

p1673639270829519-slack-C04J4UG3FLG

## Testing instructions:

* Spin up a version of this branch 
* Go to `/wp-admin/admin.php?page=manage-customers` and add a Contact
  *  You can use an existing contact as well.
* Go to the "New Quote" page (`/wp-admin/admin.php?page=zbs-add-edit&action=edit&zbstype=quote`)
* Create a new quote
* Press the `Download PDF` button at the bottom of the page
* Verify that the quote looks correct (basically: that it's generated and it isn't blank)

### CODE REVIEW

File: projects/plugins/crm/ZeroBSCRM.php [hunk 1]
```diff
@@ -23,6 +23,66 @@
...
+// NOTE: This will do for EVERY page load, so just add temporarily else adds rolling DRAIN on sys
+// define( 'ZBSPERFTEST', 1 );
+
+if ( ! defined( 'ZBS_ROOTFILE' ) ) {
+	define( 'ZBS_ROOTFILE', __FILE__ );
+	define( 'ZBS_PLUGIN_DIR', plugin_dir_path( __FILE__ ) );
+	define( 'ZBS_ROOTDIR', basename( __DIR__ ) );
+	define( 'ZBS_ROOTPLUGIN', ZBS_ROOTDIR . '/' . basename( ZBS_ROOTFILE ) );
+	define( 'ZBS_LANG_DIR', basename( __DIR__ ) . '/languages' );
+}
```
Comment (kallehauge) [inline projects/plugins/crm/ZeroBSCRM.php:36]: I've moved this further up on the page so it's defined before we try to call `ZBS_PLUGIN_DIR` for the autoloader.

File: projects/plugins/crm/.phpcs.dir.xml [hunk 2]
```diff
@@ -1,5 +1,25 @@
...
+			<property name="text_domain" value="zero-bs-crm" />
+		</properties>
+	</rule>
+
+	<rule ref="WordPress.Utils.I18nTextDomainFixer">
+		<properties>
+			<property name="old_text_domain" type="array" />
+			<property name="new_text_domain" value="zero-bs-crm" />
+		</properties>
+	</rule>
```
Comment (kallehauge) [inline projects/plugins/crm/.phpcs.dir.xml:27]: I tried to satisfy PHPCS but it's a can of worms 😂
I just stopped after this one and we'll have to merge it with the failing tests.

**Review by gogdzl — DISMISSED** [review]
Looks good to me. Thanks for taking care of this!

Original Invoice | ` update/crm/autoload-dom-pdf`'s Invoice
-|-
image | image

File: projects/plugins/crm/.phpcs.dir.xml (hunk 2 above)
Comment (anomiex) [reply projects/plugins/crm/.phpcs.dir.xml:27]: It doesn't seem that bad to me. The additions above are good to start.

I see the following other sniffs being reported on this PR:

* `Generic.WhiteSpace.ScopeIndent.IncorrectExact` - Looks like this could be fixed easily enough by indenting the function correctly. The file will be inconsistent because you have [almost](https://github.com/Automattic/jetpack/blob/cee249591c7134cb78fc19e62e1cfee0ff2a59e5/projects/plugins/crm/includes/ZeroBSCRM.SystemChecks.php#L437-L444) everything else in the file oddly indented, but at some point you'll have to fix that anyway (but I'd recommend doing that in a separate PR).
* `WordPress.Arrays.CommaAfterArrayItem.NoComma` - Easy to fix, just add the missing comma.
* `WordPress.WP.I18n.MissingTranslatorsComment` - Easy to fix, just add the missing comment..
* `WordPress.NamingConventions.ValidFunctionName.FunctionNameInvalid` - If you want to ignore this one, either with a `phpcs:disable` in ZeroBSCRM.SystemChecks.php or for your entire plugin here, I won't try to stop you. Personally I don't care for WordPress's naming conventions either, but it's the standard we've chosen to follow for Jetpack-the-plugin itself so it's part of the ruleset.

The "Check linter exclude lists" is also easy to fix, just delete the flagged line from that file.

File: projects/plugins/crm/phpcs.security.xml.dist [hunk 3]
```diff
@@ -14,7 +14,6 @@
 	<exclude-pattern>./vendor/*</exclude-pattern>
```
Comment (anomiex) [inline projects/plugins/crm/phpcs.security.xml.dist:14]: FYI, I think this file is no longer needed. PHPCS is done at the monorepo level now.

File: projects/plugins/crm/phpcs.xml.dist [hunk 4]
```diff
@@ -14,7 +14,6 @@
 	<exclude-pattern>./vendor/*</exclude-pattern>
```
Comment (anomiex) [inline projects/plugins/crm/phpcs.xml.dist:14]: FYI, I think this file is no longer needed. PHPCS is done at the monorepo level now.

**Review by anomiex — COMMENTED** [review]
Looks like most of the CI errors here are due to a bug in automattic/jetpack-composer-plugin, which will be fixed by #28369.

See inline for comments on the phpcs errors.

File: projects/plugins/crm/includes/ZeroBSCRM.SystemChecks.php [hunk 5]
```diff
@@ -25,27 +25,69 @@
...
-		$featList = array('zlib','dompdf','mb_internal_encoding','pdffonts','curl','phpver','wordpressver','locale','assetdir','executiontime','memorylimit','postmaxsize','uploadmaxfilesize','wpuploadmaxfilesize','dbver','dalver','corever','local','localtime','serverdefaulttime','sqlrights','devmode','permalinks','mysql','innodb', 'fontinstalled', 'encryptionmethod' );
+/**
+ * Check if a PHP systems/library is correctly installed.
+ *
+ * @param string $key The feature we want to know the status off.
+ * @param bool   $with_info Weather we should return a message or not.
+ *
+ * @return array|false
+ */
+function zeroBSCRM_checkSystemFeat( $key = '', $with_info = false ) {
```
Comment (kallehauge) [inline projects/plugins/crm/includes/ZeroBSCRM.SystemChecks.php:36]: Renaming this one is a can of worms because it's used in other paid functions as well and, linting wise, this will force me to update the places its being used as well which then triggers other linting issues that are outside the scope of this PR.

File: projects/plugins/crm/.phpcs.dir.xml (hunk 2 above)
Comment (kallehauge) [reply projects/plugins/crm/.phpcs.dir.xml:27]: I've fixed most of the linting issues but the "can of worms" I was mostly worried about remains (renaming `zeroBSCRM_checkSystemFeat`).

I feel like any more changes in this PR will make it too bloated as well

File: projects/plugins/crm/includes/ZeroBSCRM.SystemChecks.php (hunk 5 above)
Comment (anomiex) [reply projects/plugins/crm/includes/ZeroBSCRM.SystemChecks.php:36]: As I mentioned, I don't mind ignoring that one here. You have several options:

* Put `// phpcs:ignore WordPress.NamingConventions.ValidFunctionName.FunctionNameInvalid` at the end of this line.
* Put `// phpcs:disable WordPress.NamingConventions.ValidFunctionName.FunctionNameInvalid` near the top of the file, maybe with additional commentary as to why (e.g. permanently ignoring, or "TODO should be fixed later").
* Add something like
  ```
  <!-- We don't like these parts of the WordPress Coding Standards for this plugin. -->
  <rule ref="WordPress.NamingConventions.ValidFunctionName.FunctionNameInvalid">
  	<severity>0</severity>
  </rule>
  ```
  to `projects/plugins/crm/.phpcs.dir.xml` to ignore it for the whole plugin.

File: projects/plugins/crm/phpcs.xml.dist (hunk 4 above)
Comment (kallehauge) [reply projects/plugins/crm/phpcs.xml.dist:14]: Yeah, this one doesn't seem to provide much value anymore. I'll remove it in a separate PR - thank you!

File: projects/plugins/crm/phpcs.security.xml.dist (hunk 3 above)
Comment (kallehauge) [reply projects/plugins/crm/phpcs.security.xml.dist:14]: Yeah, that makes sense. I guess we just need to remember to copy:

```
<rule ref="WordPress.Security.EscapeOutput">
    <properties>
        <property name="customAutoEscapedFunctions" type="array" value="zbsLink,jpcrm_esc_link"/>
    </properties>
</rule>
```

... to `projects/plugins/crm/.phpcs.dir.xml` but that's about it. I'll do it in a separate PR - thank you!

**Review by gogdzl — APPROVED** [review]
Looks good to me!

**Review by anomiex — APPROVED** [review]
